File: iosApp/iosApp/Views/StateIndicators.swift
/**
 * Session state indicators for the Termtastic iOS app.
 *
 * Renders the per-session status shown next to titles in `TreeView` rows and
 * the `TerminalScreen` navigation bar, mirroring the web client's
 * `.tt-status-dot` indicator (`applyDotState()` in
 * `web/.../WebStateActions.kt`). Per issue #38 the indicator is painted in the
 * theme's foreground text colour (`Palette.textPrimary`) — not a fixed
 * green/red — so it meshes with any theme:
 *  - idle (`nil`) → a solid dot, no pulse.
 *  - `"working"`  → the same dot, breathing (pulsing) between full and ~30%.
 *  - `"waiting"`  → the dot is swapped for a pulsing warning/exclamation
 *    triangle, so working and waiting stay distinguishable now that they share
 *    one colour.
 */

import SwiftUI

/// Per-row status indicator, mirroring the web `.tt-status-dot` (issue #38): a
/// small bead whose motion/shape — but no longer its colour — encodes the
/// session state. It is always painted in the theme's foreground text colour
/// (`Palette.textPrimary`) so it meshes with any theme instead of a fixed
/// green/red.
///  - idle (`nil`) → a solid dot, no pulse.
///  - `"working"`  → the same dot, breathing between full and ~30% opacity.
///  - `"waiting"`  → a pulsing warning/exclamation triangle in place of the dot.
///
/// Rendered at the LEADING edge of each `TreeView` leaf row, aggregated on tab
/// headers, and shown next to the `TerminalScreen` title. A soft `.shadow` glow
/// surrounds the dot, echoing the web bead's box-shadow halo.
struct StatusDot: View {
    /// The session state (`"working"`, `"waiting"`, or nil/idle).
    let state: String?
    /// Square footprint in points; the core bead is ~44% of it, the rest is
    /// glow headroom.
    var box: CGFloat = 16

    /// Drives the working "breathe"; a single flip is enough for
    /// `repeatForever(autoreverses:)` to oscillate forever. Kept in sync with
    /// `state` via `.onAppear` / `.onChange` so it tracks state that changes
    /// while the row is already on screen.
    @State private var pulse = false

    var body: some View {
        let color = Palette.textPrimary
        let pulsing = state == "working"
        Group {
            if state == "waiting" {
                // Waiting for input → swap the dot for the warning triangle.
                WaitingWarningIcon(box: box, color: color)
            } else {
                Circle()
                    .fill(color)
                    .frame(width: box * 0.44, height: box * 0.44)
                    .shadow(color: color.opacity(0.7), radius: box * 0.26)
                    .frame(width: box, height: box)
                    // Web `.tt-status-dot.state-working` breathes 1 → 0.3 → 1
                    // over a 2.5s ease-in-out cycle; the 1.25s autoreversing
                    // half-cycle matches.
                    .opacity(pulsing && pulse ? 0.3 : 1.0)
                    .animation(
                        pulsing
                            ? .easeInOut(duration: 1.25).repeatForever(autoreverses: true)
                            : .default,
                        value: pulse
                    )
                    .onAppear { pulse = pulsing }
                    .onChange(of: pulsing) { _, value in pulse = value }
                    .accessibilityLabel(
                        state == "working" ? "Status: working" : "Status: idle"
                    )
            }
        }
    }
}

/// Warning triangle with an exclamation mark, fading between full and 30%
/// opacity to flag a session waiting for user input.
///
/// Geometry mirrors the web client's `.tt-status-dot.state-waiting` mask (a
/// 16-unit viewBox: a filled triangle with the exclamation bar and dot punched
/// out as transparent holes via an even-odd fill), and the fade mirrors the
/// `fade-warning` keyframes (opacity 1 → 0.3 → 1 over 2.5s ease-in-out). Drawn
/// in the supplied `color` (the theme foreground) so it meshes with any theme.
private struct WaitingWarningIcon: View {
    /// Square footprint in points (matches the dot it replaces).
    let box: CGFloat
    /// Fill colour (the theme's foreground text colour).
    let color: Color
    /// Drives the repeat-forever fade; flipped once on appear.
    @State private var pulse = false

    var body: some View {
        Canvas { context, canvasSize in
            // Scale the 16-unit viewBox to fill the canvas; the triangle has
            // its own padding (apex at y=1.7, base at y=13.7) so it never
            // touches the edges.
            let s = min(canvasSize.width, canvasSize.height) / 16
            var path = Path()
            // Triangle silhouette.
            path.move(to: CGPoint(x: 8 * s, y: 1.7 * s))
            path.addLine(to: CGPoint(x: 14.5 * s, y: 13.7 * s))
            path.addLine(to: CGPoint(x: 1.5 * s, y: 13.7 * s))
            path.closeSubpath()
            // Exclamation bar (cut-out).
            path.addRect(CGRect(x: 7.25 * s, y: 5.6 * s, width: 1.5 * s, height: 4.2 * s))
            // Exclamation dot (cut-out).
            path.addEllipse(in: CGRect(x: 7.1 * s, y: 10.9 * s, width: 1.8 * s, height: 1.8 * s))
            // Even-odd fill turns the bar + dot into transparent holes, so the
            // exclamation reads on any theme background.
            context.fill(path, with: .color(color), style: FillStyle(eoFill: true))
        }
        .frame(width: box, height: box)
        .opacity(pulse ? 0.3 : 1.0)
        .animation(.easeInOut(duration: 1.25).repeatForever(autoreverses: true), value: pulse)
        .onAppear { pulse = true }
        .accessibilityLabel("Status: waiting for input")
    }
}

File: androidApp/src/main/kotlin/se/soderbjorn/termtastic/android/ui/StateIndicators.kt
/**
 * Session state indicators for the Termtastic Android app.
 *
 * Renders the per-session status shown next to titles in [TreeScreen] rows and
 * the [TerminalScreen] top bar, mirroring the web client's `.tt-status-dot`
 * indicator (`applyDotState()` in `web/.../WebStateActions.kt`). Per issue #38
 * the indicator is painted in the theme's foreground text colour (not a fixed
 * green/red) so it meshes with any theme:
 *  - idle (`null`) → a solid dot, no pulse.
 *  - `"working"`   → the same dot, breathing (pulsing) between full and ~30%.
 *  - `"waiting"`   → the dot is swapped for a pulsing warning/exclamation
 *    triangle, so working and waiting stay distinguishable now that they share
 *    one colour.
 *
 * @see TreeScreen
 * @see TerminalScreen
 */
package se.soderbjorn.termtastic.android.ui

import androidx.compose.animation.core.CubicBezierEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathFillType
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.stateDescription
import androidx.compose.ui.unit.dp

/** CSS `ease-in-out` curve, matching the web `fade-warning` animation timing. */
private val EaseInOutCss = CubicBezierEasing(0.42f, 0f, 0.58f, 1f)

/**
 * Per-row status indicator mirroring the web `.tt-status-dot` (issue #38): a
 * small bead whose motion/shape — but no longer its colour — encodes the
 * session state. It is always painted in the theme's foreground text colour
 * ([SidebarTextPrimary]) so it meshes with any theme instead of a fixed
 * green/red.
 *  - idle (`null`) → a solid dot, no pulse.
 *  - `"working"`   → the same dot, breathing between full and ~30% opacity.
 *  - `"waiting"`   → a pulsing warning/exclamation triangle (see
 *    [WaitingWarningIcon]) in place of the dot.
 *
 * Rendered at the LEADING edge of each [TreeScreen] leaf row and aggregated on
 * tab headers; also shown next to the [TerminalScreen] title. A soft
 * radial-gradient glow surrounds the dot, echoing the web bead's box-shadow
 * halo.
 *
 * @param state the session state (`"working"`, `"waiting"`, or null/idle).
 * @param boxDp the square canvas size in dp; the core bead is ~44% of it, the
 *   rest is glow headroom.
 * @see WaitingWarningIcon
 */
@Composable
internal fun StatusDot(state: String?, boxDp: Int = 16) {
    val color = SidebarTextPrimary
    // Waiting for input → swap the dot for the warning triangle, still pulsing.
    if (state == "waiting") {
        WaitingWarningIcon(boxDp = boxDp, color = color)
        return
    }
    val pulsing = state == "working"
    val alpha = if (pulsing) {
        val transition = rememberInfiniteTransition(label = "statusDot")
        transition.animateFloat(
            initialValue = 1f,
            targetValue = 0.3f,
            animationSpec = infiniteRepeatable(
                // The web `.tt-status-dot.state-working` breathes over a 2.5s
                // ease-in-out cycle; the reversing tween runs the 1250ms
                // half-cycle each way so it matches.
                animation = tween(durationMillis = 1250, easing = EaseInOutCss),
                repeatMode = RepeatMode.Reverse,
            ),
            label = "statusDotAlpha",
        ).value
    } else {
        1f
    }
    val desc = if (state == "working") "working" else "idle"
    Canvas(
        modifier = Modifier
            .size(boxDp.dp)
            .semantics { stateDescription = desc },
    ) {
        val center = Offset(size.width / 2f, size.height / 2f)
        val core = size.minDimension * 0.22f
        val glow = size.minDimension * 0.5f
        // Soft outer glow — a radial gradient fading to transparent at the
        // canvas edge, echoing the web dot's layered box-shadow halo.
        drawCircle(
            brush = Brush.radialGradient(
                colors = listOf(color.copy(alpha = 0.5f * alpha), Color.Transparent),
                center = center,
                radius = glow,
            ),
            radius = glow,
            center = center,
        )
        // Core bead.
        drawCircle(color = color.copy(alpha = alpha), radius = core, center = center)
    }
}

/**
 * Warning triangle with an exclamation mark, fading between full and 30%
 * opacity to flag a session waiting for user input.
 *
 * Geometry mirrors the web client's `.tt-status-dot.state-waiting` mask (a
 * 16-unit viewBox: a filled triangle with the exclamation bar and dot punched
 * out as transparent holes via an even-odd fill), and the fade mirrors the
 * `fade-warning` keyframes (opacity 1 → 0.3 → 1 over 2.5s ease-in-out). Painted
 * in the supplied [color] (the theme foreground) so it meshes with any theme.
 *
 * @param boxDp the square canvas size in dp (matches the dot it replaces).
 * @param color the fill colour (the theme's foreground text colour).
 * @see StatusDot
 */
@Composable
private fun WaitingWarningIcon(boxDp: Int, color: Color) {
    val transition = rememberInfiniteTransition(label = "fadeWarning")
    val alpha = transition.animateFloat(
        initialValue = 1f,
        targetValue = 0.3f,
        animationSpec = infiniteRepeatable(
            animation = tween(durationMillis = 1250, easing = EaseInOutCss),
            repeatMode = RepeatMode.Reverse,
        ),
        label = "fadeWarningAlpha",
    ).value
    Canvas(
        modifier = Modifier
            .size(boxDp.dp)
            .semantics { stateDescription = "waiting for input" },
    ) {
        // Scale the 16-unit viewBox to fill the canvas; the triangle has its
        // own padding (apex at y=1.7, base at y=13.7) so it never touches the
        // edges. A single even-odd path fills the triangle and cuts the
        // exclamation bar + dot as holes (transparent, reading on any theme).
        val s = size.minDimension / 16f
        val path = Path().apply {
            fillType = PathFillType.EvenOdd
            // Triangle silhouette.
            moveTo(8f * s, 1.7f * s)
            lineTo(14.5f * s, 13.7f * s)
            lineTo(1.5f * s, 13.7f * s)
            close()
            // Exclamation bar (cut-out).
            addRect(Rect(7.25f * s, 5.6f * s, 8.75f * s, 9.8f * s))
            // Exclamation dot (cut-out).
            addOval(Rect(7.1f * s, 10.9f * s, 8.9f * s, 12.7f * s))
        }
        drawPath(path, color = color.copy(alpha = color.alpha * alpha))
    }
}
